src/Microsoft.Azure.ServiceBus/MessageBatchReceivePump.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Core;
+    using Primitives;
+
+    sealed class MessageBatchReceivePump
+    {
+        readonly Func<IList<Message>, CancellationToken, Task> onMessageBatchCallback;
+        readonly string endpoint;
+        readonly MessageBatchHandlerOptions registerHandlerOptions;
+        readonly MessageReceiver messageReceiver;
+        readonly CancellationToken pumpCancellationToken;
+        readonly SemaphoreSlim maxConcurrentCallsSemaphoreSlim;
+        readonly ServiceBusDiagnosticSource diagnosticSource;
+
+        public MessageBatchReceivePump(MessageReceiver messageReceiver,
+            MessageBatchHandlerOptions registerHandlerOptions,
+            Func<IList<Message>, CancellationToken, Task> callback,
+            Uri endpoint,
+            CancellationToken pumpCancellationToken)
+        {
+            this.messageReceiver = messageReceiver ?? throw new ArgumentNullException(nameof(messageReceiver));
+            this.registerHandlerOptions = registerHandlerOptions;
+            this.onMessageBatchCallback = callback;
+            this.endpoint = endpoint.Authority;
+            this.pumpCancellationToken = pumpCancellationToken;
+            this.maxConcurrentCallsSemaphoreSlim = new SemaphoreSlim(this.registerHandlerOptions.MaxConcurrentCalls);
+            this.diagnosticSource = new ServiceBusDiagnosticSource(messageReceiver.Path, endpoint);
+        }
+
+        public void StartPump()
+        {
+            TaskExtensionHelper.Schedule(() => this.MessageBatchPumpTaskAsync());
+        }
+
+        bool ShouldRenewLock()
+        {
+            return
+                this.messageReceiver.ReceiveMode == ReceiveMode.PeekLock &&
+                this.registerHandlerOptions.AutoRenewLock;
+        }
+
+        Task RaiseExceptionReceived(Exception e, string action)
+        {
+            var eventArgs = new ExceptionReceivedEventArgs(e, action, this.endpoint, this.messageReceiver.Path, this.messageReceiver.ClientId);
+            return this.registerHandlerOptions.RaiseExceptionReceived(eventArgs);
+        }
+
+        async Task MessageBatchPumpTaskAsync()
+        {
+            while (!this.pumpCancellationToken.IsCancellationRequested)
+            {
+                IList<Message> messages = null;
+                try
+                {
+                    await this.maxConcurrentCallsSemaphoreSlim.WaitAsync(this.pumpCancellationToken).ConfigureAwait(false);
+                    messages = await this.messageReceiver.ReceiveAsync(this.registerHandlerOptions.MaxMessageCount, this.registerHandlerOptions.ReceiveTimeOut).ConfigureAwait(false);
+
```

Comment:
Sorry to break this to you, but receiveing in a batch still doesn't work as expected and its a very high possibliity that you will end up with a batch size of 1 always. It needs some changes across client and service

Reply:
I havn't done that much testing but I've tried sending 10 000 messages, max message count of 500 with prefetch 1000. I received 300-500 messages each pump. Can look into it more deeply though when I get the time.